File: Common/src/main/java/dev/common/constant/ApiConstant.java
package dev.common.constant;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;

public class ApiConstant {
    public static final class AUTHENTICATION_URL {
        public static final String SERVICE_NAME = "AUTHENTICATION";
        public static final String URL = "/api/v1/authentication";
        public static final String REGISTER = "/register";
        public static final String AUTHENTICATE = "/authenticate";
        public static final String EMPLOYEE_AUTHENTICATION = "/authentication-employee";
        public static final String EMPLOYEE = "/create-employee";
        public static final String EXCHANGE_TOKEN = "/exchange-token";
        public static final String LOGOUT = "/logout";
        public static final class INTERNAL_URL{
            public static final String URL = "/api/v1/internal/authentication";
            public static final String CHECK_PHONE_NUMBER_EXIST = "/check-phone-number";
        }
    }
    public static final class HOSPITAL_INFORMATION {
        public static final String SERVICE_NAME = "HOSPITALINFORMATION";
        public static final String PROVINCE_URL = "/api/v1/provinces";
        public static final String CHECK_ADDRESS = "/check-address";
        public static final String EXAMINATION_ROOM_URL = "/api/v1/examination-rooms";
        public static final String CHECK_ROOM_EXIST = "/check-room/{id}";
        public static final String ID = "{id}";
        public static final String GET_BY_IDS = "/find-by-ids";
        public static final String CHECK_DISEASES_EXIST = "/check";
        public static final String DISEASE_URL = "/api/v1/diseases";
    }

    public static final class FACULTY_URL{
        public static final String URL = "/api/v1/faculties";
        public static final String ID = "/{id}";
        public static final String POSITION_URL = "/api/v1/positions";
        public static final String FACULTY_ID = "/faculty/{id}";
    }

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static final class EmployeeUrl {
        public static final String URL = "/api/v1/employees";
        public static final String SERVICE_NAME = "EMPLOYEE";
        public static final String ROLE_URL = "/api/v1/roles";
        public static final String ID = "/{id}";
        public static final String FIND_BY_IDS = "/find-by-ids";
        public static final String GET_LOGGED_USER_INFORMATION = "/get-logged-user-information";
    }

    public static final class GREETING_URL{
        public static final String EXAMINATION_FORM_URL = "/api/v1/greeting";
        public static final String TICKET_URL = "/api/v1/tickets";
        public static final String PRINT = "/print";
        public static final String FIRST_TIME = "/first-time";
        public static final String WITH_APPOINTMENT = "/with-appointment";
        public static final String FIND_RECEIVED_PATIENTS_TODAY = "find/received-patients-today";
    }

    public static final class WORKING_SCHEDULE_URL{
        public static final String URL = "/api/v1/working-schedules";
        public static final String SERVICE_NAME = "WORKINGSCHEDULE";
        public static final String ID = "/{id}";
        public static final String SEARCH = "/search";
        public static final String CHECK_SCHEDULE_TODAY = "/check-schedule-today/{id}";
        public static final String GET_SCHEDULE_TODAY_OF_EMPLOYEE = "/today/{id}";
        public static final String GET_SCHEDULES_IN_MONTH_OF_EMPLOYEE = "/schedules-in-month";
        public static final String GET_SCHEDULES_BY_DATE = "/find/by-date";
    }

    public static final class EXAMINATION_RESULT_URL{
        public static final String RESULT_URL = "/api/examination-results";
        public static final String ID = "/{id}";
        public static final String CONSULTATION_FORM = "/api/consultation-form";

        public static final String APPOINTMENT_FORM_NEXT_EXAMINATION = "/api/v1/appointment-form-next-examination";
        public static final String FIND_WAITING_EXAMINATION_PATIENTS = "/find/waiting-examination-patients";
    }

    public static final class MEDICINE_URL{
        public static final String SERVICE_NAME = "MEDICINE";
        public static final String MEDICINE_URL = "/api/v1/medicines";
        public static final String CREATE_PATIENT_MEDICINE_INVOICE = "/create-patient-medicine-invoice";
        public static final String ID = "/{id}";
        public static final String CHECK_MEDICINES_EXIST = "/check";
        public static final String ORIGIN_URL = "/api/v1/origins";
        public static final String SUPPLIER_URL = "/api/v1/suppliers";
        public static final String IMPORT_INVOICE_URL = "/api/v1/import-invoices";
        public static final String SEARCH = "/search";
    }

    public static final class PATIENT{
        public static final String SERVICE_NAME = "PATIENT";
        public static final String PATIENT_URL = "/api/patients";
        public static final String GET_BY_IDS = "/get/by-ids";
        public static final String ID = "/{id}";
        public static final String CHECK_EXIST_PATIENT = "/check/{id}";
        public static final String APPOINTMENT_URL = "/api/appointments";
        public static final String GET_APPOINTMENTS_OF_TODAY = "/today";
        public static final String GET_EXAMINED_APPOINTMENTS_OF_TODAY = "/today/examined";
        public static final String FIND_DETAILS_BY_APPOINTMENT_ID = "/find-detail-by-appointment-id/{id}";
        public static final String DOCTOR_CREATE_APPOINTMENT = "/create-appointment-by-doctor";
        public static final String GET_LOGGED_USER_INFORMATION = "/get-logged-user-information";
    }

    public static final class PAYMENT{
        public static final String EXAMINATION_COST_URL = "/api/examination-costs";
        public static final String ID = "/{id}";

        public static final String INVOICE_URL = "/api/invoices";
    }

    public static final class CHAT{
        public static final String CHAT_URL = "/api/chat";
        public static final String RECEIVER_ID = "/with/{id}";
        public static final String SEND_IMAGE = "/send-image/{id}";
        public static final String RELATION_SHIP_URL = "/api/relation-ship";
        public static final String FIND_RELATION_SHIPS_OF_DOCTOR = "/find-relation-ships-of-doctor";
        public static final String FIND_RELATION_SHIPS_OF_PATIENT = "/find-relation-ships-of-patient";
    }
}

File: FrontEnd/src/ApiConstant.js

class HOST{
    static getHost = () => { return "http://localhost:9010"}
}

class MinioUrl{
    static getHost = () => {
        return "http://localhost:9000"
    }

    static downloadFile = (url) => {
        return this.getHost() + `/${url}`
    }
}

class AUTHENTICATION{
    static getUrl = () => {
        return HOST.getHost() + "/api/v1/authentication";
    }

    static authenticate = () => {
        return this.getUrl() + "/authenticate";
    }

    static exchangeToken = () => {
        return this.getUrl() + "/exchange-token";
    }

    static authenticateEmployee = () => {
        return this.getUrl() + "/authentication-employee"
    }

    static createEmployee = () => {
        return this.getUrl() + "/create-employee"
    }

    static logout = () => {
        return this.getUrl() + "/logout"
    }
}

class EMPLOYYEE{
    static getUrl = (permission) => {
        if(permission === null || permission === '')
            return HOST.getHost() + '/api/v1/employees'
        return HOST.getHost() + `/api/v1/employees?role=${permission}`
    }

    static id = (id) => {
        return this.getUrl() + "/" + id;
    }

    static findByIds = () => {
        return this.getUrl('') + "/find-by-ids"
    }

    static getLoggedUserInformation = () => {
        return this.getUrl('') + "/get-logged-user-information"
    }
}

class MEDICINE{
    static Medicine = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/v1/medicines";
        }

        static getById = (id) => {
            return this.getUrl() + `/${id}`;
        }

        static search = (q) => {
            return this.getUrl() + `/search?q=${q}`
        }

        static getByIDs = () =>{
            return this.getUrl() + '/get-by-ids'
        }
    }

    static Origin = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/v1/origins";
        }
    }
}

class ROLE{
    static role = new Map([
        ['ADMIN', "Quản trị viên"],
        ['DOCTOR', "Bác sĩ"],
        ['RECEPTION_STAFF', "Nhân viên tiếp đón"],
        ['MEDICINE_DISPENSER', "Nhân viên kho thuốc"]
    ])

    static getRole = (role) => {
        return this.role.get(role)
    }

    static getUrl = () => {
        return HOST.getHost() + "/api/v1/roles";
    }
}

class WEBSOCKET{
    static getUrl = () => {
        return  "http://localhost:9010/hospital_system";
    }

    static topicCreateEmployee = (id) => {
        return "/topic/create-employee/" + id;
    }

    static topicCreatedEmployee = (id) => {
        return "/topic/created-employee/" + id;
    }

    static processedImage = (id) => {
        return "/topic/processed-image/" + id
    }

    static updatedNumberExaminationForm = (id) => {
        return "/topic/updated-number-examination-form/" + id
    }

    static chat = (id) => {
        return `/topic/chat/user/${id}`
    }

    static appSendImage = (id) => {
        return `/app/chat/image/doctor/${id}`
    }
}

class HOSPITAL_INFORMATION{
    static EXAMINATION_ROOM = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/examination-rooms"
        }

        static findByIds = () => {
            return this.getUrl() + "/find-by-ids"
        }
    }
    static DISEASES = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/diseases"
        }
    }

    static ADDRESS = class{
        static getUrl = () => {
            return HOST.getHost() + "/api/provinces"
        }

        static getByProvinceId = (id) => {
            return this.getUrl() + `/by-province/${id}`
        }

        static getByDistrictId = (id) => {
            return this.getUrl() + `/by-district/${id}`
        }

        static getDetail = () => {
            return this.getUrl() + "/get-address-detail"
        }
    }
}

class WORKING_SCHEDULE{
    static getUrl = () => {
        return HOST.getHost() + "/api/v1/working-schedules"
    }

    static getSchedulesInMonthOfEmployee = (year, month) => {
        return this.getUrl() + `/schedules-in-month?year=${year}&month=${month}`
    }

    static getUrlById = (id) => {
        return this.getUrl() + "/" + id;
    }

    static getSchedulesByDate = (date) => {
        return this.getUrl() + `/find/by-date?date=${date}`
    }
}

class AI{
    static getUrl = () => {
        return HOST.getHost() + "/api/v1/ai"
    }

    static diagnostic = () => {
        return this.getUrl() + "/diagnostic"
    }
}

class PATIENT{
    static APPOINTMENT = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/appointments"
        }

        static getAppointmentsOfToday = () => {
            return this.getUrl() + "/today"
        }

        static findAppointmentDetailByAppointmentId = (id) => {
            return this.getUrl() + "/find-detail-by-appointment-id/" + id
        }

        static doctorCreateAppointment = () => {
            return this.getUrl() + "/create-appointment-by-doctor"
        }

        static getAppointmentsOfPatient = (year, month) => {
            return this.getUrl() + `/get-appointments-of-patient?year=${year}&month=${month}`
        }

        static getById = (id) => {
            return this.getUrl() + "/" + id
        }
    }

    static PATIENT_API = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/patients"
        }

        static getByIds = () => {
            return this.getUrl() + "/get/by-ids"
        }

        static byId = (id) => {
            return this.getUrl() + "/" + id
        }

        static getUserInformation = () =>{
            return this.getUrl() + "/get-logged-user-information";
        }
    }
}


class Greeting{
    static ExaminationForm = class{
        static getGetUrl = () => {
            return HOST.getHost() + "/api/v1/greeting"
        }

        static withAppointment = () => {
            return this.getGetUrl() + "/with-appointment"
        }

        static findReceivedPatientsToday = () => {
            return this.getGetUrl() + "/find/received-patients-today";
        }

        static firstTime = () => {
            return this.getGetUrl() + "/first-time"
        }
    }
}

class ExaminationResult{
    static ExaminationResultUrl = class{
        static getUrl = () => {
            return HOST.getHost() + "/api/examination-results"
        }

        static findWaitingExaminationPatients = () => {
            return this.getUrl() + '/find/waiting-examination-patients'
        }

        static findById = (id) => {
            return this.getUrl() + "/" + id
        }

        static findHistoriesOfPatient = () => {
            return this.getUrl() + "/find-histories-of-patient"
        }

        static findHistoriesByPatientId = (id) => {
            return this.getUrl() + `/find-by-patient-id/${id}`
        }

        static getExaminedResultTodayOfDoctor = () => {
            return this.getUrl() + "/get-examined-result-today-of-doctor"
        }
    }

    static MedicineConsultationForm = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/consultation-form"
        }

        static byId = (id) => {
            return this.getUrl() + `/${id}`
        }
    }
}

const Chat = class {
    static Message = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/chat"
        }

        static getMessageByReceiverId = (id, pageable) => {
            return this.getUrl() + `/with/${id}?page=${pageable.page}`
        }

        static sendImage = (id) => {
            return this.getUrl() + "/send-image/" + id
        }
    }

    static RelationShip = class {
        static getUrl = () => {
            return HOST.getHost() + "/api/relation-ship"
        }

        static findRelationShipsOfPatient = (page) => {
            return this.getUrl() + `/find-relation-ships-of-patient?page=${page}`
        }

        static findRelationShipOfDoctor = (page) => {
            return this.getUrl() + `/find-relation-ships-of-doctor?page=${page}`
        }
    }
}

const PaymentApi = class {
    static getUrl = () => {
        return HOST.getHost() + "/api/invoices"
    }

    static getUnPaid = () => {
        return this.getUrl() + "/unpaid"
    }

    static byId = (id) => {
        return this.getUrl() + `/${id}`
    }

    static payInCash = (id) => {
        return this.getUrl() + `/pay-in-cash/${id}`
    }

    static payByVnpay = (id) => {
        return this.getUrl() + `/pay-by-vnpay/${id}`
    }

    static EXAMINATION_COST_URL = () => {
        return HOST.getHost() + "/api/examination-costs"
    }
}

const VNPAY = class {
    static errorCode = new Map([
        ['00', "Giao dịch thành công."],
        ['07', "Trừ tiền thành công. Giao dịch bị nghi ngờ (liên quan tới lừa đảo, giao dịch bất thường)."],
        ['09', "Giao dịch không thành công do: Thẻ/Tài khoản của khách hàng chưa đăng ký dịch vụ InternetBanking tại ngân hàng."],
        ['10', "Giao dịch không thành công do: Khách hàng xác thực thông tin thẻ/tài khoản không đúng quá 3 lần"],
        ['11', 'Giao dịch không thành công do: Đã hết hạn chờ thanh toán. Xin quý khách vui lòng thực hiện lại giao dịch.'],
        ['12', 'Giao dịch không thành công do: Thẻ/Tài khoản của khách hàng bị khóa.'],
        ['13', 'Giao dịch không thành công do Quý khách nhập sai mật khẩu xác thực giao dịch (OTP). Xin quý khách vui lòng thực hiện lại giao dịch.'],
        ['24', 'Giao dịch không thành công do: Khách hàng hủy giao dịch'],
        ['51', 'Giao dịch không thành công do: Tài khoản của quý khách không đủ số dư để thực hiện giao dịch.'],
        ['65', 'Giao dịch không thành công do: Tài khoản của Quý khách đã vượt quá hạn mức giao dịch trong ngày.'],
        ['75', 'Ngân hàng thanh toán đang bảo trì.'],
        ['79', 'Giao dịch không thành công do: KH nhập sai mật khẩu thanh toán quá số lần quy định. Xin quý khách vui lòng thực hiện lại giao dịch'],
        ['99', 'Các lỗi khác (lỗi còn lại, không có trong danh sách mã lỗi đã liệt kê)']
    ])
}

export {HOST, AUTHENTICATION, EMPLOYYEE, ROLE, WEBSOCKET, MEDICINE, HOSPITAL_INFORMATION, WORKING_SCHEDULE, AI, PATIENT, Greeting, ExaminationResult, Chat, MinioUrl, PaymentApi,
VNPAY}